### DISCUSSION TITLE
Simulating thermal contraction in fractured systems using MultiApps

### DISCUSSION THREAD

**Discussion by WilkAndy** [idaholab/moose#18257] [Q&A Modules: General] [answered]
Hi everyone,

I'm canvassing ideas on how to simulate thermal contraction in fractured systems using MultiApps.  This extending some work I've been doing with @rpodgorney , @lynnmunday , @cbolisetti and KT on solving non-isothermal flow through fractured porous media.  My question takes a few pictures to understand....

Imagine a solid material (the "matrix") that contains a thin plate-like void (the "fracture").  Let's just think of the 2D case, where the matrix is a 2D square and the fracture is a 1D line segment:

![fracture_matrix](https://user-images.githubusercontent.com/6676821/124425658-36e4a680-ddac-11eb-912e-8d8b67849d2b.png)

(fracture is the red line)

Imagine injecting cold fluid into the fracture.  After some time, the cold will diffuse a little way into the matrix.  This cold "sheath" around the fracture will thermally contract, leading to an increase in the fracture aperture:

![fracture_matrix_cold](https://user-images.githubusercontent.com/6676821/124426067-dc981580-ddac-11eb-8292-d58755255f0b.png)

In the work I've been doing, I've been using a MultiApp approach: one App models the fracture physics and one App models the matrix physics.  One big reason for doing this is that realistic fractures are actually complicated networks that are difficult to mesh, and extremely difficult to mesh as part of a 3D solid system, so keeping the fracture mesh separate from the solid mesh is very advantageous.  Another advantage is computational speed, since the fracture physics is often quite complicated (not just heat flow, even though that's what I'm concentrating on in this Discussion).  In this approach, heat energy from the fracture App appears in the matrix App as a set of Dirac sinks/sources.

![matrix_app](https://user-images.githubusercontent.com/6676821/124426699-bf177b80-ddad-11eb-81f0-fa1bf6f09cf5.png)

(matrix elements are shown as squares, fracture Dirac sinks/sources are red dots)

**My end game is to accurately predict fracture aperture as a function of temperature in the system.**

However, a straightforward thermal contraction in the matrix is exactly the opposite of what I want to see: the cold elements contract, pulling the remaining elements towards them.  This is because the matrix mesh does not contain an explicit fracture:

![matrix_app_cold](https://user-images.githubusercontent.com/6676821/124427029-32b98880-ddae-11eb-81b3-abacc59f6c42.png)

(matrix elements have contracted due to cold Dirac sources)

So, how can I accurately predict fracture aperture using this MultiApp approach?

I can think of a few not-so-good ways of doing this (using an analytic solution, cutting matrix elements, empirical approach).  Can you think of some potential ways forward?  I'm not asking for "the one-and-only answer" (although that'd be great, heheheh) - I'm just hoping someone might initiate a good approach.

**Comment by hugary1995** [comment]
Here is an easy hack if your domain is symmetric. Utilize symmetry to only model the left half of the domain. The symmetry condition is simply u_x = 0 on the symmetry plane. Now, to represent a discontinuity, just don't prescribe Dirichlet BC on the part of the symmetry plane that overlaps with the fracture -- leave it stress-free. If other physics are modeled correctly, I expect to see crack opening.

**Reply by hugary1995** [reply]
If the injecting fluid is pressurized, you can also apply a traction using the same approach. 

If the fluid is invading into the porous matrix, then I recommend you to use a phase field to model single phase flow into a poroelastic media on the matrix mesh. But it seems to me this isn't something you are considering as of now.

**Reply by WilkAndy** [reply]
Thanks @hugary1995 - those are great answers.  Unfortunately, the situation is more complicated than the above pictures: there are many elliptical fracture planes intersecting each other all over the place.  I just drew the simple situation to make the problem clearer.  Yes, also we have fluid in the fracture and in the matrix, and those physics are governed by PorousFlow, and geomechanics too, but all those things can be handled easily enough.  The thermal contraction and its impact on the fracture aperture is the difficult bit.  (The same problematic phenomenon also occurs for mechanical deformation from fluid pressure, but i thought thermal contraction/expansion was easier to explain.)

**Reply by hugary1995** [reply]
Okay, If it's arbitrary smooth cracks then Xfem is probably your best bet.

If you know the crack locations a priori and they are not propagating, then the czm capabilities in moose would also be useful.

If I were solving this problem from the very beginning, I probably would use phase field. It can handle hydraulic fracture in porous media, with Darcy flow, stokes flow, etc. There are plenty of literature on this subject. Several papers by Mary Wheeler, Chad Landis, and Blaise Bourdin would be a good starting point. But it seems that the commitments so far prevent you from choosing a completely different method.

**Reply by WilkAndy** [reply]
Thanks @hugary1995 - really useful comments.

**Reply by rpodgorney** [reply]
Can't this be done simply by tracking the displacement in the matrix, adjacent and normal to the fracture plane, and then use that displacement to essentially calculate an assumed change in aperture?  Record keeping would likely be messy, but you are already looking at normals for the heat/mass transfers, correct?

**Reply by jiangwen84** [reply]
I might give the exact the same suggestions as @hugary1995 mentioned. However, I have seen the the examples from @lynnmunday where you have extremely complicated fracture network. At this moment, neither XFEM nor CZM can work nicely. 

It seems the issue comes from the fact that the matrix does not model the fracture explicitly. Instead of adding the sink source term, could you degrade the matrix stiffness at fracture locations? It is kind of like the smeared damage approach in some sense.

**Reply by WilkAndy** [reply]
> I might give the exact the same suggestions as @hugary1995 mentioned. However, I have seen the the examples from @lynnmunday where you have extremely complicated fracture network. At this moment, neither XFEM nor CZM can work nicely.
> 
> It seems the issue comes from the fact that the matrix does not model the fracture explicitly. Instead of adding the sink source term, could you degrade the matrix stiffness at fracture locations? It is kind of like the smeared damage approach in some sense.

Ha, it seems we reached the same conclusion about degrading the fracture stiffness, @jiangwen84 (see below - for some reason i didn't see your post until i had written mine).

**Reply by WilkAndy** [reply]
> Can't this be done simply by tracking the displacement in the matrix, adjacent and normal to the fracture plane, and then use that displacement to essentially calculate an assumed change in aperture? Record keeping would likely be messy, but you are already looking at normals for the heat/mass transfers, correct?

I'm not sure how that works @rpodgorney .   The matrix deformation is the opposite of what we want - see the last picture above.

**Comment by WilkAndy** [comment, marked as answer]
After mulling over this a little further, I'm wondering whether "the best" approach might be to use appropriate elasticity tensor for the matrix elements containing the fractures.  In this case E_{ijxx}=0, or more generally E_{ijmn}P_{mk}{P_{nl}, where P_{mk}=delta_{mk}-n_{m}n_{k}, is the projection operator and n is the normal to the fracture plane.  (Also could use Cosserat with fracture normal stiffness = 0, and suitable fracture shear stiffness.)  This will mean that any heat load will cause zero stress_{xx}, so the element won't contract along the x direction.  As the cool temperature hits the next layer of matrix elements, they will contract, pulling the fracture-containing elements with them, leading to the correct result.  This approach is only spatially accurate to the size of one matrix element, but that's exactly the accuracy we get using the MultiApp approach anyway.

**Reply by WilkAndy** [reply]
Note to future self: If i decide to go down this path then
E_ijxx = 0 means "any strain_xx will not generate any stress"
E_xxij = 0 means "any strain will not generate any stress_xx"
Both of these are probably what i want.

But perhaps I also want E_ijkx = E_ijxl = 0, meaning "a shear-strain involving x will generate no stress", and E_xjkl = E_ixkl = 0, meaning "any strain will not generate any shear-stress involving x".  Obviously if the fracture had zero shear strength, 2 of these would be correct (E_ijkx = 0 = E_xjkl), but the other 2 follow only by symmetry of E and are not necessarily physically correct.

Cosserat mechanics gets around this by having non-symmetric stress, strain and E, so can have E_xjkl = 0 but E_ixkl != 0.   But the disadvantage is a considerable increase in compute intensity for little gain: Cosserat is actually only needed in the few matrix elements that contain fractures, and an estimate of fracture shear strength is needed, and Cosserat is only true in the small layer-thickness limit anyway.

So I can't see any clear right or wrong in selecting E_ijkx = E_ijxl = E_xjkl = E_ixkl = 0.  (Or more generally E_mnop P_{mi}P_{nj}P_{ok}P_{pl}.)  My intuition tells me it is better to use these because I want to treat these matrix elements like very flexible material.

**Reply by hugary1995** [reply]
I've done this before -- see this writeup:
![Screenshot from 2021-07-06 20-51-44](https://user-images.githubusercontent.com/31555204/124683882-1562e800-de9c-11eb-8584-9ca8ff83909d.png)

**Reply by hugary1995** [reply]
In your case \Gamma-convergence follows from h-refinement.

**Reply by WilkAndy** [reply]
Yea, that sort of thing, @hugary1995 .   Interesting that the author of that material agreed with me about setting the shear stresses to "zero", but i don't see why that's always correct - where did the above material come from?

In my case i won't have to worry about crack closing, i believe.  (Cracks might close, but the whole point of this is to find the crack aperture, and if it becomes negative i'll just say aperture=zero (actually, probably just a small number like the initial crack aperture).)   So i won't have the numerically-annoying discontinuity at strain=0.

**Reply by hugary1995** [reply]
I wrote it just for you :)

I agree, your case is simpler.

**Reply by WilkAndy** [reply]
Another node:  have to be aware of potential problems when multiple fractures sit within one matrix element, and when fractures sit in neighbouring matrix elements.

**Comment by rpodgorney** [comment]
I was envisioning using the difference, and only allowing the aperture to increase---but that won’t work.

I like your thoughts on the elasticity approach!


--
Robert K. Podgorney
Chief Computational Scientist
Advanced Scientific Computing Division
Idaho National Laboratory

Office: 208-526-1524<tel:208-526-1524>
Cell: 208-520-9361<tel:208-520-9361>
email: ***@***.******@***.***>
Statement of Confidentiality
The contents of this e-mail message and any attachments are confidential and are intended solely for addressee. The information may also be legally privileged. This transmission is sent in trust, for the sole purpose of delivery to the intended recipient. If you have received this transmission in error, any use, reproduction or dissemination of this transmission is strictly prohibited. If you are not the intended recipient, please immediately notify the sender by reply e-mail or phone and delete this message and its attachments, if any.


From: Andy Wilkins ***@***.***>
Reply-To: idaholab/moose ***@***.***>
Date: Tuesday, July 6, 2021 at 3:59 PM
To: idaholab/moose ***@***.***>
Cc: Robert Podgorney ***@***.***>, Mention ***@***.***>
Subject: [EXTERNAL] Re: [idaholab/moose] Simulating thermal contraction in fractured systems using MultiApps (#18257)


Can't this be done simply by tracking the displacement in the matrix, adjacent and normal to the fracture plane, and then use that displacement to essentially calculate an assumed change in aperture? Record keeping would likely be messy, but you are already looking at normals for the heat/mass transfers, correct?

I'm not sure how that works @rpodgorney<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Frpodgorney&data=04%7C01%7Crobert.podgorney%40inl.gov%7C966aaa403d9447481d6808d940c9425e%7C4cf464b7869a42368da2a98566485554%7C0%7C0%7C637612055410784626%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=nJXWyOiF%2BWxHacjVGJ2uQdRqg6kia13zaEA3nKIkjn8%3D&reserved=0> . The matrix deformation is the opposite of what we want - see the last picture above.

**Reply by lynnmunday** [reply]
@WilkAndy I think the fracture planes would be far enough apart that an analytic solution for thermal contraction of a half plane on the fracture sub-app should accurately capture the aperture change.

**Reply by WilkAndy** [reply]
Yea, i agree that in many cases that will be true, @lynnmunday

**Comment by rpodgorney** [comment]
@lynnmunday — this is probably true, and would be fast—but I would like to explore a full mechanical solve option.  I know we’ll get questions from my technical review board

@wilkandy — if you’d like to explore further and have the time we can increase the scope/budget on the contract.

I am really excited to see the progress you guys are making, well done!



-- sent from my mobile device, please excuse any typos or autocorrect errors


Robert K. Podgorney, PhD

Chief Computational Scientist

Advanced Scientific Computing Division

Idaho National Laboratory

Office: 208-526-1524<tel:208-526-1524>

Cell: 208-520-9361<tel:208-520-9361>

email: ***@***.******@***.***>

Statement of Confidentiality

The contents of this e-mail message and any attachments are confidential and are intended solely for addressee. The information may also be legally privileged. This transmission is sent in trust, for the sole purpose of delivery to the intended recipient. If you have received this transmission in error, any use, reproduction or dissemination of this transmission is strictly prohibited. If you are not the intended recipient, please immediately notify the sender by reply e-mail or phone and delete this message and its attachments, if any.

________________________________
From: Lynn Munday ***@***.***>
Sent: Wednesday, July 14, 2021 5:14:31 PM
To: idaholab/moose ***@***.***>
Cc: Robert K. Podgorney ***@***.***>; Mention ***@***.***>
Subject: [EXTERNAL] Re: [idaholab/moose] Simulating thermal contraction in fractured systems using MultiApps (#18257)


@WilkAndy<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FWilkAndy&data=04%7C01%7Crobert.podgorney%40inl.gov%7Cccc0384eb7594364320c08d9472584ec%7C4cf464b7869a42368da2a98566485554%7C0%7C0%7C637619048752584021%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=VqRBPrVo4WY38%2FhNrFzvT%2F%2FivvRPoKQkEXuXuM7Yzn0%3D&reserved=0> I think the fracture planes would be far enough apart that an analytic solution for thermal contraction of a half plane on the fracture sub-app should accurately capture the aperture change.
